arrow-fx-coroutines/src/test/kotlin/arrow/fx/coroutines/stream/ops/CallbackTest.kt

```diff
@@ -0,0 +1,235 @@
+      val effect = Atomic(false)
+      val s = Stream.callback<Int> {
+        effect.set(true)
+        end()
+      }
+
+      effect.get() shouldBe false
+      s.compile().drain()
+      effect.get() shouldBe true
+    }
+  }
+
+  "emits values" {
+    checkAll(Arb.list(Arb.int())) { list ->
+      Stream.callback {
+        list.forEach { emit(it) }
+        end()
+      }
+        .compile()
+        .toList() shouldBe list
+    }
+  }
+
+  "emits varargs" {
+    checkAll(Arb.list(Arb.int()).map { it.toTypedArray() }) { list ->
+      Stream.callback {
+        emit(*list)
+        end()
+      }
+        .chunks()
+        .compile()
+        .toList() shouldBe listOf(Chunk(*list))
+    }
+  }
+
+  "emits iterable" {
+    checkAll(Arb.list(Arb.int())) { list ->
+      Stream.callback<Int> {
+        emit(list)
+        end()
+      }
+        .chunks()
+        .compile()
+        .toList() shouldBe listOf(Chunk.iterable(list))
+    }
+  }
+
+  "emits chunks" {
+    checkAll(Arb.chunk(Arb.int()), Arb.chunk(Arb.int())) { ch, ch2 ->
+      Stream.callback<Int> {
+        emit(ch)
+        emit(ch2)
+        end()
+      }
+        .chunks()
+        .compile()
+        .toList() shouldBe listOf(ch, ch2)
+    }
+  }
+
+  "long running emission" {
+    Stream.callback {
+      ForkAndForget {
+        countToCallback(4, { it }, { emit(it) }) { end() }
+      }
+    }
+      .compile()
+      .toList() shouldBe listOf(1, 2, 3, 4, 5)
+  }
+
+  "parallel emission/pulling" {
+    val ref = Atomic(false)
+
+    Stream.callback {
+      emit(1)
+      sleep(1.seconds)
+      emit(2)
+      ref.set(true)
+      end()
+    }
+      .effectMap {
+        if (it == 1) ref.get() shouldBe false
+        else Unit
+      }
+      .compile()
+      .drain()
+  }
+
+  "emits and completes" {
+    Stream.callback {
+      emit(1)
+      emit(2)
+      emit(3)
+      end()
+    }
+      .compile()
+      .toList() shouldBe listOf(1, 2, 3)
+  }
+
+  "forwards errors" {
+    checkAll(Arb.throwable()) { e ->
+      val s = Stream.cancellableCallback<Int> {
+        throw e
+      }
+        .compile()
+
+      assertThrowable {
+        s.drain()
+      } shouldBe e
+    }
+  }
+
+  "forwards suspended errors" {
+    checkAll(Arb.throwable()) { e ->
+      val s = Stream.cancellableCallback<Int> {
+        e.suspend()
+      }
+        .compile()
+
+      assertThrowable {
+        s.drain()
+      } shouldBe e
+    }
+  }
+
+  "runs cancel token" {
+    checkAll(Arb.int()) {
+      val latch = Promise<Unit>()
+      val effect = Atomic(false)
+
+      val s = Stream.cancellableCallback<Int> {
+        CancelToken { effect.set(true) }
+      }
+
+      val f = ForkAndForget {
+        parTupledN(
+          { s.compile().drain() },
+          { latch.complete(Unit) }
+        )
+      }
+
+      parTupledN({ latch.get() }, { sleep(20.milliseconds) })
+
+      f.cancel()
+
+      effect.get() shouldBe true
+    }
+  }
+
+  "can cancel never" {
```

Comment:
We would need to refactor this test somehow, any idea? cc @nomisRev

Reply:
Hey @aballano,
Sorry for the long wait :/ I aligned the semantics of `Stream.cancellable` to `cancellable` as discussed, and also tried to update the documentation and examples to clarify how these functions work.